packages/vm/test/api/EIPs/eip-2935-historical-block-hashes.spec.ts

```diff
@@ -0,0 +1,167 @@
+import { Blockchain } from '@ethereumjs/blockchain'
+import { Common, Hardfork } from '@ethereumjs/common'
+import {
+  Address,
+  BIGINT_1,
+  bigIntToBytes,
+  bigIntToHex,
+  equalsBytes,
+  setLengthLeft,
+  zeros,
+} from '@ethereumjs/util'
+import { hexToBytes } from 'ethereum-cryptography/utils'
+import { assert, describe, it } from 'vitest'
+
+import { VM } from '../../../src/vm'
+
+function eip2935ActiveAtCommon(timestamp: number) {
+  const hfs = [
+    Hardfork.Chainstart,
+    Hardfork.Homestead,
+    Hardfork.Dao,
+    Hardfork.Chainstart,
+    Hardfork.SpuriousDragon,
+    Hardfork.Byzantium,
+    Hardfork.Constantinople,
+    Hardfork.Petersburg,
+    Hardfork.Istanbul,
+    Hardfork.Berlin,
+    Hardfork.London,
+    Hardfork.ArrowGlacier,
+    Hardfork.GrayGlacier,
+    Hardfork.Shanghai,
+    Hardfork.Paris,
+    Hardfork.MergeForkIdTransition,
+    Hardfork.Shanghai,
+    Hardfork.Cancun,
+  ]
+  const hardforks = []
+  for (const hf of hfs) {
+    hardforks.push({
+      name: hf,
+      block: 0,
+    })
+  }
+  hardforks.push({
+    name: 'testEIP2935Hardfork',
+    block: null,
+    timestamp,
+  })
+  const c = Common.custom({
+    customHardforks: {
+      testEIP2935Hardfork: {
+        name: 'testEIP2935Hardfork',
+        comment: 'Start of the Ethereum main chain',
+        url: '',
+        status: 'final',
+        eips: [2935],
+      },
+    },
+    hardforks,
+    /*genesis: {
+      gasLimit: 30_000_000,
+      timestamp: "0x0",
+      extraData: "0x",
+      difficulty: "0x0",
+      nonce: "0x0000000000000000"
+    }*/
+  })
+  return c
+}
+
+const commonGenesis = eip2935ActiveAtCommon(1)
+commonGenesis.setHardforkBy({
+  timestamp: 1,
+})
+const historyAddress = Address.fromString(
+  bigIntToHex(commonGenesis.param('vm', 'historyStorageAddress'))
+)
+
+describe('EIP 2935: historical block hashes', () => {
+  it('should save genesis block hash to the history block hash contract', async () => {
+    commonGenesis.setHardfork(Hardfork.Chainstart)
+    const blockchain = await Blockchain.create({
+      common: commonGenesis,
+      validateBlocks: false,
+      validateConsensus: false,
+    })
+    const vm = await VM.create({ common: commonGenesis, blockchain })
+    commonGenesis.setHardforkBy({
+      timestamp: 1,
+    })
+    const genesis = await vm.blockchain.getBlock(0)
+    const block = await (
+      await vm.buildBlock({
+        parentBlock: genesis,
+        blockOpts: {
+          putBlockIntoBlockchain: false,
+        },
+      })
+    ).build()
+    await vm.blockchain.putBlock(block)
+    await vm.runBlock({ block, generate: true })
+
+    const storage = await vm.stateManager.getContractStorage(
+      historyAddress,
+      setLengthLeft(bigIntToBytes(BigInt(0)), 32)
+    )
+    assert.ok(equalsBytes(storage, genesis.hash()))
+  })
+  it('should ensure blocks older than 256 blocks can be retrieved from the history contract', async () => {
```

Comment:
@holgerd77 had to re-check, but this test *should* actually test for the off-by-one error.